### DISCUSSION TITLE
Unable to execute "server create_db"

### DISCUSSION THREAD

**Discussion by willy-dauphin** [getredash/redash#6357] [Q&A]
Hi, 
I'm not longer able to install Redash using Docker. This is my (new) configuration : 

- Ubuntu : 20.04
- Redash : 10.1.0
- Node : 16.0.0
- Yarn : 1.22.19
- Docker : 24.0.5
- Docker-compose : 1.25.0

Everything works fine until this commande : "docker-compose run --rm server create_db". And this is what appends :

`root@xxx-test:~/redash# docker-compose run --rm server create_db
Starting redash_redis_1 ... done
Starting redash_postgres_1 ... done
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/event/base.py", line 98, in __getattr__
    ls = self._empty_listeners[name]
KeyError: 'instrument_class'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/manage.py", line 6, in <module>
    from redash.cli import manager
  File "/app/redash/__init__.py", line 53, in <module>
    import_query_runners(settings.QUERY_RUNNERS)
  File "/app/redash/query_runner/__init__.py", line 436, in import_query_runners
    __import__(runner_import)
  File "/app/redash/query_runner/query_results.py", line 5, in <module>
    from redash import models
  File "/app/redash/models/__init__.py", line 31, in <module>
    from redash.models.base import (
  File "/app/redash/models/base.py", line 39, in <module>
    make_searchable(db.metadata, options={"regconfig": "pg_catalog.simple"})
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy_searchable/__init__.py", line 563, in make_searchable
    mapper, 'instrument_class', manager.process_mapper
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/event/api.py", line 102, in listen
    _event_key(target, identifier, fn).listen(*args, **kw)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/event/registry.py", line 204, in listen
    dispatch_collection = getattr(target.dispatch, identifier)
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/event/base.py", line 100, in __getattr__
    raise AttributeError(name)
AttributeError: instrument_class`

Any idea about this problem ?
Thanks in advance for your help :)

**Comment by justinclift** [comment]
Are you trying to install Docker for development purposes, or trying to install Docker as part of a production deployment?

If it's for development purposes, please use this wiki page for your local development setup as it's known to work with our current source code: :smile:

https://github.com/getredash/redash/wiki/Local-development-setup

**Reply by willy-dauphin** [reply]
Thanks for your answer. I'm trying to install Docker as part of a production deployment : another tip ? :)

**Reply by justinclift** [reply]
Are you using the setup.sh script here?

https://github.com/getredash/setup/

**Reply by willy-dauphin** [reply]
No, I'm following this : https://redash.io/help/open-source/dev-guide/docker
Unfortunately, the database creation failed.

How can I use the setup.sh ? Any tutorial ? Thanks again.

**Reply by justinclift** [reply]
That's a "fairly simple" (!) automatic setup script, targeted towards being run on Ubuntu 20.04 LTS.

My (strong) suggestion would be to first experiment on a (new) test system, running Ubuntu 20.04 LTS.

On such a system, you should be able to download the "setup.sh" script by itself:

https://github.com/getredash/setup/raw/master/setup.sh

And run that on the system.  That script will install the needed pieces (including Docker), then install the last official Redash release.

In theory (!) that should work ok.  After the install then Redash will be running, with the "Initial Redash Setup" page being displayed on the web interface of the server.

Does that help? :smile:

**Reply by willy-dauphin** [reply]
Yes, a lot ! Thanks again for yout help.

Now, I've a new error : 
`sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "postgres" (XXX.XXX.XXX.X), port 5432 failed: FATAL:  password authentication failed for user "postgres"

(Background on this error at: http://sqlalche.me/e/e3q8)
`
This is the postgres configuration setup.sh : 
`  COOKIE_SECRET=$(pwgen -1s 32)
  SECRET_KEY=$(pwgen -1s 32)
  POSTGRES_PASSWORD=$(pwgen -1s 32)
  REDASH_DATABASE_URL="postgresql://postgres:$POSTGRES_PASSWORD@postgres/postgres"
`

This is the postgres configuration from docker-compose.yml : 
`
  postgres:
    image: postgres:13-alpine
    env_file: /opt/redash/env
    volumes:
      - /root/opt/postgres-data:/var/lib/postgresql/data
    restart: always
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 10s
      timeout: 5s
      retries: 5
`

Any idea ? :)

**Reply by justinclift** [reply]
Ahhh.  Yeah, that `FATAL: password authentication failed for user "postgres"` error when you're newly setting things up *generally* means you've run the script more than once, and it's overwritten the correct PostgreSQL password in the `/opt/redash/env` file.

The solution will be to shut down the docker containers:

```
cd /opt/redash
docker-compose down
```

Then delete the `/opt/redash` directory completely and run the script again.  Once. :smile:

**Reply by willy-dauphin** [reply]
Amazing ! Thanks a lot ! 
According to you, what's the best option to edit a docker-compose.yml before running the script ?

**Reply by justinclift** [reply]
Hmmm, I'd probably do something like this:

1. Make a fork of that setup repository (eg to github.com/willy-dauphin/setup)
2. Change the file(s) you want in that repository
3. Change the setup.sh script (it's just a text file) so it grabs the docker-compose.yml from your repository instead of the getredash/setup one.

That should work.  In theory. :smile:

**Reply by willy-dauphin** [reply]
Thanks again ! It works with a local personnal docker-compose.yml configuration file. Unfortunately, I still have a postgres error.

Error : 
`sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) connection to server at "postgres" (XXX.XX.X.X), port 5432 failed: FATAL:  password authentication failed for user "postgres"`

I tried as you said : 
cd redash/ 
docker-compose down
cd ..
rm -r redash/
./setup/setup.sh

env file : 
```
PYTHONUNBUFFERED=0
REDASH_LOG_LEVEL=INFO
REDASH_REDIS_URL=redis://redis:6379/0
POSTGRES_PASSWORD=AAA
REDASH_COOKIE_SECRET=XXX
REDASH_SECRET_KEY=YYY
REDASH_DATABASE_URL=postgresql://postgres:AAA@postgres/postgres
REDASH_RATELIMIT_ENABLED=false
REDASH_ENFORCE_CSRF=true
```
But unfortunately, nothing seems to work. Any other idea ?

**Reply by willy-dauphin** [reply]
I've tried it with an existing database and it works. The problem seems to be with the _"sudo docker-compose run --rm server create_db"_ command. Any idea ?

**Reply by justinclift** [reply]
Interesting.  That sounds (to me), like the docker containers weren't fully shut down before removing the `/opt/redash` directory.  That could result in some part of the old database still hanging around.

That being said, this isn't a total blocker.  There's some mucking around involved, but you *can* manually log into the PostgreSQL container on the unix level, then reset the password for the `postgres` user in the database there.  That should at least allow things to work.

I'll demonstrate below, using my local development setup.  For reference, when checking the docker services running at the moment they are:

```
$ docker-compose ps
       Name                     Command                  State                   Ports             
---------------------------------------------------------------------------------------------------
redash_email_1       bin/maildev                      Up (healthy)   0.0.0.0:1025->1025/tcp,:::1025
                                                                     ->1025/tcp, 0.0.0.0:1080->1080
                                                                     /tcp,:::1080->1080/tcp        
redash_postgres_1    /usr/local/bin/docker-entr ...   Up             0.0.0.0:15432->5432/tcp,:::154
                                                                     32->5432/tcp                  
redash_redis_1       docker-entrypoint.sh redis ...   Up             6379/tcp                      
redash_scheduler_1   /app/bin/docker-entrypoint ...   Up             5000/tcp                      
redash_server_1      /app/bin/docker-entrypoint ...   Up             0.0.0.0:5001->5000/tcp,:::5001
                                                                     ->5000/tcp, 0.0.0.0:5678->5678
                                                                     /tcp,:::5678->5678/tcp        
redash_worker_1      /app/bin/docker-entrypoint ...   Up             5000/tcp
```

To change the password of the `postgres` user in the PostgreSQL container, we only really need the `redash_postgres_1` container to be running.  Having the others "Up" won't hurt through.

First, we jump into the PostgreSQL container as the `postgres` unix user:

```
$ docker-compose exec -u postgres postgres bash
c0bb85df8b21:~$
```

The `c0bb85df8b21` in the returned prompt is just the docker internal container name.  Yours will be different.

From there, start the PostgreSQL command line interface ("psql"), then issue the SQL command for changing the user password:

```
c0bb85df8b21:~$ psql postgres
psql (15.3)
Type "help" for help.

postgres=# ALTER USER postgres WITH PASSWORD 'AAA';
ALTER ROLE
postgres=# quit
c0bb85df8b21:~$ exit
```

Make sure to use the same password there as you do in the `env` file (`POSTGRES_PASSWORD`, `REDASH_DATABASE_URL`).

You *might* need to restart the other containers again (not sure):

```
$ docker-compose restart
Restarting redash_scheduler_1 ... done
Restarting redash_worker_1    ... done
Restarting redash_server_1    ... done
Restarting redash_email_1     ... done
Restarting redash_postgres_1  ... done
Restarting redash_redis_1     ... done
```

Anyway, try that and see how to go. :smile:

**Reply by willy-dauphin** [reply]
Thanks a lot for your help ! It's now working perfectly ! 😄 
A last question : with this setup, how can I run an equivalent of "yarn build" ? Indeed, I would like to edit the ColorPalette.ts file.

**Reply by justinclift** [reply]
Hmmm, that's a bit more involved.  You could run commands for adding a custom `ColorPalette.ts` in say the `server` container.  But the changes won't be persisted to disk.  So, they'll disappear when you restart the container.

You *could* probably achieve the result by building a new Redash server container, basing it upon the existing production release.  You'd need to create a new `Dockerfile` and tell it what commands you want run on the container.

It's not super involved, so you could figure it out in a few hours of mucking around if that's your kind of thing?

**Reply by willy-dauphin** [reply]
Is it impossible to configure a "volume" file in the "server" container configuration from my _docker-compose.yml_ ? I would like to use a custom file. It seems working because when I'm looking inside the container, the file has been changed. Unfortunately, there is no appearant changes in the application :/

**Reply by justinclift** [reply]
Ahhh.  If you just want to override a specific file, that's indeed doable using the `volume` approach.

For example, lets say your base installation directly is `/opt/redash`, you can create an override like this:

```
    volumes:
      - /opt/redash/overrides/dynamic_settings.py:/app/redash/settings/dynamic_settings.py:ro
```

That overrides the `/app/redash/settings/dynamic_settings.py` file in a container with the custom version stored in `/opt/redash/overrides/dynamic_settings.py` on the host.  The `:ro` text fragment on the end tells Docker it's Read Only, for safety. :smile:

Btw, that `overrides` directory in `/opt/redash` above is just for neatness, so I can have several different overrides stored in one spot for easy management.

So, you'd make an override for `ColorPalette.ts` using the same approach.

With that change in your `docker-compose.yml`, then apply it using:

```
$ cd /opt/redash
$ docker-compose up -d
```

**Comment by justinclift** [comment]
Oh.  I just realised something.

Since you're changing a .ts file, then you might indeed need to rebuild the whole front end part of things for that to become effective.

So, overriding that single file might make no difference at all, as seems to be what happens for you. :frowning:

**Reply by willy-dauphin** [reply]
Unfortunately, it's right. Any idea to rebuild the whole front end ?

**Reply by justinclift** [reply]
I have an idea, but it's completely untested. :wink:

You'll need to create your own Docker image, based on the production image, and with the changes you want added in.

It shouldn't be super difficult though, and it looks like you're willing to put in some time and effort. :smile:

---

### 1. First step is to make a copy of the Redash repository somewhere.  Definitely *not* in the `/opt/redash` directory.

```
$ mkdir ~/git_repos
$ cd ~/git_repos
$ git clone https://github.com/getredash/redash
```

### 2. Roll back the source code to the Redash 10.1.0 production release

```
$ cd redash
$ git reset --hard v10.1.0
```

### 3. Replace the files you want changed (eg `ColorPalette.ts`)

### 4. Replace the existing `Dockerfile` in the repository with something like this:

```
FROM node:16.20.1 as frontend-builder

RUN npm install --global --force yarn@1.22.19

RUN useradd -m -d /frontend redash
USER redash

WORKDIR /frontend
COPY --chown=redash package.json yarn.lock .yarnrc /frontend/
COPY --chown=redash viz-lib /frontend/viz-lib

RUN yarn --frozen-lockfile --network-concurrency 1

COPY --chown=redash client /frontend/client
COPY --chown=redash webpack.config.js /frontend/
RUN yarn build

FROM redash/redash:10.1.0.b50633

WORKDIR /app

COPY --chown=redash . /app
COPY --from=frontend-builder --chown=redash /frontend/client/dist /app/client/dist
RUN chown redash /app
USER redash

ENTRYPOINT ["/app/bin/docker-entrypoint"]
CMD ["server"] 
```

### 5. Build the image

```
$ docker build .
```

I say "something like" that above, as that Dockerfile stuff is just my assembling things by eyeball without actual testing.  You'll probably need to hammer it into the right shape a bit more.

That's mostly copied from [here](https://github.com/getredash/redash/blob/master/Dockerfile).  It includes the front end building stuff which makes sense, but uses the production Redash image to supply the backend part.

Are you ok to try the above?  Jumping on our Discord channel might be useful while doing this, if that's your kind of thing. :smile:

**Reply by willy-dauphin** [reply]
Wow ! Thanks a lot for your complete answer. It's an amazing idea ! I'll discuss on it with my team and see if we are able / if we have time to do it. Just another idea : is it impossible to simply "yarn build" inside our server-container ? :)

**Reply by justinclift** [reply]
> is it impossible to simply "yarn build" inside our server-container ?

Very unlikely to work.  The docker images are built in a two step process, with the front end stuff being built in a temporary container which has all the needed tools (npm, yarn, etc).

The output from that step (but without the tools themselves) gets copied across to a fresh container with the Python backend pieces in it, and that's what makes the final Redash image.

So, `yarn` and the other front end development pieces aren't in the server container image.  You *could* install them and do things in there, but that's pretty much just doing the same thing as the above Dockerfile approach. :smile:
